### ISSUE TITLE
Log details as one message

### ISSUE THREAD

**Issue by robachmann** [OpenFeign/feign#874]
**tl;dr:** Call `log()` only once for each request and response?

I'm using `feign.client.config.default.loggerLevel: full` to log payload of an application deployed to Cloud Foundry in a dev/test environment. In Cloud Foundry, every line of log to `stdout` is treated as an independent log-event and sent to a log-drain. Already with little load, those events may get out of order and scatter eventually in the target system. 

As I see in the code, Feign logs every detail separately on a new line: https://github.com/OpenFeign/feign/blob/master/core/src/main/java/feign/Logger.java#L46

I'm wondering if I can trick it to log everything as a single event where new lines are simply rendered with '\n' (which then can be [processed by logback](https://github.com/cloudfoundry/loggregator-release/blob/develop/docs/java-multi-line-work-around.md) to have it on a single line) or if there could be an extension that generates the log-message in a first step and log that message with a single call to `log()` afterwards?

**Comment by kdavisk6** [comment]
You can achieve this by providing your own `Logger` implementation.  If you had one that buffered each log call and only echo when you want, probably in the `logAndRebufferResponse()` method.  That's called after the response is complete.  Thoughts?

**Comment by robachmann** [comment]
Thank you Kevin for your thoughts. I fear, however, that this solution might not be thread-safe. If I understand you correctly, you're suggesting that logs are aggregated over a certain time and printed in at the end of `logAndRebufferResponse`. I assume this would mean I'd have to maintain that list of loggable lines in my custom logger. My poor man's solution is currently to simply overwrite the specific methods of `feign.Logger` as follows to produce a single string:

```
@Slf4j
public class CustomLogger extends feign.Logger {

    @Override
    protected void logRequest(String configKey, Level logLevel, Request request) {
        if (log.isDebugEnabled()) {
            List<String> logs = new ArrayList<>();

            logs.add(format(configKey, "---> %s %s HTTP/1.1", request.method(), request.url()));
            if (logLevel.ordinal() >= Level.HEADERS.ordinal()) {

                for (String field : request.headers().keySet()) {
                    for (String value : valuesOrEmpty(request.headers(), field)) {
                        logs.add(format(configKey, "%s: %s", field, value));
                    }
                }

                int bodyLength = 0;
                if (request.body() != null) {
                    bodyLength = request.body().length;
                    if (logLevel.ordinal() >= Level.FULL.ordinal()) {
                        String
                                bodyText =
                                request.charset() != null ? new String(request.body(), request.charset()) : null;
                        logs.add(format(configKey, "")); // CRLF
                        logs.add(format(configKey, "%s", bodyText != null ? bodyText : "Binary data"));
                    }
                }
                logs.add(format(configKey, "---> END HTTP (%s-byte body)", bodyLength));
            }
            log.debug(String.join("\n", logs));
        }
    }

    @Override
    protected Response logAndRebufferResponse(String configKey, Level logLevel, Response response,
                                              long elapsedTime) throws IOException {
        if (log.isDebugEnabled()) {
            List<String> logs = new ArrayList<>();

            String reason = response.reason() != null && logLevel.compareTo(Level.NONE) > 0 ?
                    " " + response.reason() : "";
            int status = response.status();
            logs.add(format(configKey, "<--- HTTP/1.1 %s%s (%sms)", status, reason, elapsedTime));
            if (logLevel.ordinal() >= Level.HEADERS.ordinal()) {

                for (String field : response.headers().keySet()) {
                    for (String value : valuesOrEmpty(response.headers(), field)) {
                        logs.add(format(configKey, "%s: %s", field, value));
                    }
                }

                int bodyLength = 0;
                if (response.body() != null && !(status == 204 || status == 205)) {
                    // HTTP 204 No Content "...response MUST NOT include a message-body"
                    // HTTP 205 Reset Content "...response MUST NOT include an entity"
                    if (logLevel.ordinal() >= Level.FULL.ordinal()) {
                        logs.add(format(configKey, "")); // CRLF
                    }
                    byte[] bodyData = Util.toByteArray(response.body().asInputStream());
                    bodyLength = bodyData.length;
                    if (logLevel.ordinal() >= Level.FULL.ordinal() && bodyLength > 0) {
                        logs.add(format(configKey, "%s", decodeOrDefault(bodyData, UTF_8, "Binary data")));
                    }
                    logs.add(format(configKey, "<--- END HTTP (%s-byte body)", bodyLength));
                    log.debug(String.join("\n", logs));
                    return response.toBuilder().body(bodyData).build();
                } else {
                    logs.add(format(configKey, "<--- END HTTP (%s-byte body)", bodyLength));
                }
            }
            log.debug(String.join("\n", logs));
            return response;
        }
        return response;
    }

    @Override
    protected void log(String configKey, String format, Object... args) {
        // Not using SLF4J's support for parameterized messages (even though it would be more efficient) because it would
        // require the incoming message formats to be SLF4J-specific.
        if (log.isDebugEnabled()) {
            log.debug(format(configKey, format, args));
        }
    }

    private String format(String configKey, String format, Object... args) {
        return String.format(methodTag(configKey) + format, args);
    }
}
```

If my assumption is wrong and there is a way to identify and differentiate logs of the same request/response, I'd be interested in knowing how :-) 

Thank you again and best regards

**Comment by robachmann** [comment]
Thank you, Kevin. I'm slightly confused though because I think you understood my correctly indeed and I found a solution to do exactly that :-)

I close this issue as you showed me how to collect the log-lines of a request or a response and print all of them in one statement. 

Thank you again!

**Comment by erickreevestenorio** [comment]
> Thank you Kevin for your thoughts. I fear, however, that this solution might not be thread-safe. If I understand you correctly, you're suggesting that logs are aggregated over a certain time and printed in at the end of `logAndRebufferResponse`. I assume this would mean I'd have to maintain that list of loggable lines in my custom logger. My poor man's solution is currently to simply overwrite the specific methods of `feign.Logger` as follows to produce a single string:
> 
> ```
> @Slf4j
> public class CustomLogger extends feign.Logger {
> 
>     @Override
>     protected void logRequest(String configKey, Level logLevel, Request request) {
>         if (log.isDebugEnabled()) {
>             List<String> logs = new ArrayList<>();
> 
>             logs.add(format(configKey, "---> %s %s HTTP/1.1", request.method(), request.url()));
>             if (logLevel.ordinal() >= Level.HEADERS.ordinal()) {
> 
>                 for (String field : request.headers().keySet()) {
>                     for (String value : valuesOrEmpty(request.headers(), field)) {
>                         logs.add(format(configKey, "%s: %s", field, value));
>                     }
>                 }
> 
>                 int bodyLength = 0;
>                 if (request.body() != null) {
>                     bodyLength = request.body().length;
>                     if (logLevel.ordinal() >= Level.FULL.ordinal()) {
>                         String
>                                 bodyText =
>                                 request.charset() != null ? new String(request.body(), request.charset()) : null;
>                         logs.add(format(configKey, "")); // CRLF
>                         logs.add(format(configKey, "%s", bodyText != null ? bodyText : "Binary data"));
>                     }
>                 }
>                 logs.add(format(configKey, "---> END HTTP (%s-byte body)", bodyLength));
>             }
>             log.debug(String.join("\n", logs));
>         }
>     }
> 
>     @Override
>     protected Response logAndRebufferResponse(String configKey, Level logLevel, Response response,
>                                               long elapsedTime) throws IOException {
>         if (log.isDebugEnabled()) {
>             List<String> logs = new ArrayList<>();
> 
>             String reason = response.reason() != null && logLevel.compareTo(Level.NONE) > 0 ?
>                     " " + response.reason() : "";
>             int status = response.status();
>             logs.add(format(configKey, "<--- HTTP/1.1 %s%s (%sms)", status, reason, elapsedTime));
>             if (logLevel.ordinal() >= Level.HEADERS.ordinal()) {
> 
>                 for (String field : response.headers().keySet()) {
>                     for (String value : valuesOrEmpty(response.headers(), field)) {
>                         logs.add(format(configKey, "%s: %s", field, value));
>                     }
>                 }
> 
>                 int bodyLength = 0;
>                 if (response.body() != null && !(status == 204 || status == 205)) {
>                     // HTTP 204 No Content "...response MUST NOT include a message-body"
>                     // HTTP 205 Reset Content "...response MUST NOT include an entity"
>                     if (logLevel.ordinal() >= Level.FULL.ordinal()) {
>                         logs.add(format(configKey, "")); // CRLF
>                     }
>                     byte[] bodyData = Util.toByteArray(response.body().asInputStream());
>                     bodyLength = bodyData.length;
>                     if (logLevel.ordinal() >= Level.FULL.ordinal() && bodyLength > 0) {
>                         logs.add(format(configKey, "%s", decodeOrDefault(bodyData, UTF_8, "Binary data")));
>                     }
>                     logs.add(format(configKey, "<--- END HTTP (%s-byte body)", bodyLength));
>                     log.debug(String.join("\n", logs));
>                     return response.toBuilder().body(bodyData).build();
>                 } else {
>                     logs.add(format(configKey, "<--- END HTTP (%s-byte body)", bodyLength));
>                 }
>             }
>             log.debug(String.join("\n", logs));
>             return response;
>         }
>         return response;
>     }
> 
>     @Override
>     protected void log(String configKey, String format, Object... args) {
>         // Not using SLF4J's support for parameterized messages (even though it would be more efficient) because it would
>         // require the incoming message formats to be SLF4J-specific.
>         if (log.isDebugEnabled()) {
>             log.debug(format(configKey, format, args));
>         }
>     }
> 
>     private String format(String configKey, String format, Object... args) {
>         return String.format(methodTag(configKey) + format, args);
>     }
> }
> ```
> 
> If my assumption is wrong and there is a way to identify and differentiate logs of the same request/response, I'd be interested in knowing how :-)
> 
> Thank you again and best regards

do you have configured anything for this?
